lib/Sema/CSStep.cpp

```diff
@@ -76,6 +76,12 @@ StepResult SplitterStep::resume(bool prevFailed) {
   for (auto &component : Components)
     workList.splice(workList.end(), component);
 
+  // We want to compute the number of leaf scopes as if the system had
+  // not been split, so allow each additional component after the
+  // first to have a free leaf scope that doesn't count against the
+  // total for the system.
+  CS.decrementLeafScopes(Components.size() - 1);
```

Comment:
As far as I understand the behavior, we can only reach "leaf" scope when:

1. None of the previous components failed but current did (and components which follow it up are not going to allocate scopes);
2. Disjunction choice has been "attempted" without problems - choice constraint has been simplified, but then failed to attempt any type variables;
3. Last successful type variable in the path has been bound (because all type variables attempted before it would create intermediate scopes).

So the question is - if we always decrement by the number of components here but not all of the components resulted in at least one recorded "leaf" scope (e.g. one failed so all the rest had to be aborted), wouldn't it mean that this counter would go negative and/or overflow?

Reply:
We count something as a leaf scope any time it creates no new scopes. That could be because we found a successful solution or because a constraint failed. So we should always increment this at least once per component.

My hope here is that this counter will provide the same value whether we split a system or not. i.e., if we take a system and solve it without splitting it and hit 10 leafs, I would like to ensure that if we took the same system and split it into three we would also end up with a count of 10 leafs (rather than 10 plus two extra ones for the two successful endpoints in the two new components).

Does that make sense and if so are there still any concerns?

I can look at adding asserts to ensure we never roll over or go negative.

Reply:
Take a look at the ComponentStep::take, we don’t allocate scopes if previous components failed...